### ISSUE TITLE
Migrate All Addons to Auto-refresh

### ISSUE THREAD

**Issue by TheColaber** [ScratchAddons/ScratchAddons#1673]
tl;dr We need to make all addons not need a refresh to update.

After #1500 is merged (Which should be soon), we will have the ability to make all current theme addons that need a refresh to not have one when the addon is disabled while it is in use. Right now I think there is only two addons (One we have done already) that we need to migrate.

Secondly, after we have a PR that allows a new "late" event (An event that should be fired if the page has already been loaded and an addon was enabled "late"), this should allow all existing themes that need a refresh to be enabled no longer need one.

Finally, After a PR is created and merged for #783, we now have the ability for all addons to no longer need a refresh to update their state. I think from then on we would really want any new addon PRs to include these new events and variables. For example, if a userscript was added "late", we would have something such as `addon.self.late`, which would be a boolean. Some addons might need to do something different depending on if they were late or not. Addons that have already been added should try to get migrated too.

**Comment by Weredime** [comment]
`default-project` won't need auto-refresh as it works only once (which is the projects creation flow) and if we cancel before default-project is finished it may cause errors

**Comment by WorldLanguages** [comment]
> `default-project` won't need auto-refresh as it works only once (which is the projects creation flow)

What about File>New?

**Comment by Weredime** [comment]
> > `default-project` won't need auto-refresh as it works only once (which is the projects creation flow)
> 
> What about File>New?

Forgot about that

**Comment by Weredime** [comment]
Quick script I wrote  that calculates how many addons we need to migrate
```js
const fs = require("fs");

const addons = require("./addons/addons.json");

const fullRefresh = [];
const partialRefreshEnable = [];
const partialRefreshDisable = [];
const noRefresh = [];
for (const addon of addons) {
    if (addon.startsWith("//")) continue;
    const manifest = JSON.parse(
        fs.readFileSync(`addons/${addon}/addon.json`)
    );
    if (manifest.dynamicEnable && manifest.dynamicDisable) {
        fullRefresh.push(addon);
    } else if (manifest.dynamicEnable) {
        partialRefreshEnable.push(addon);
    } else if (manifest.dynamicDisable) {
        partialRefreshDisable.push(addon);
    } else {
        noRefresh.push(addon);
    };
}
for (const ref of fullRefresh) {
    console.log(`✅ ${ref} (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/${ref})`);
}
for (const ref of partialRefreshEnable) {
    console.log(`🔄 (enable) ${ref} (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/${ref})`);
}
for (const ref of partialRefreshDisable) {
    console.log(`🔄 (disable) ${ref} (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/${ref})`);
}

for (const ref of noRefresh) {
    console.log(`❌ ${ref} (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/${ref})`);
}

console.log(`--End of output--`)
console.log(`${Math.floor((fullRefresh.length / addons.length) * 100)}% (${fullRefresh.length}/${addons.length}) of addons are migrated to full refresh.\n${Math.floor((partialRefreshEnable.length / addons.length) * 100)}% (${partialRefreshEnable.length}/${addons.length}) of addons are partial refresh (enable).\n${Math.floor((partialRefreshDisable.length / addons.length) * 100)}% (${partialRefreshDisable.length}/${addons.length}) of addons are partial refresh (disable).\n${Math.floor((noRefresh.length / addons.length) * 100)}% (${noRefresh.length}/${addons.length}) of addons are not migrated to refresh.`);
```
```
✅ editor-searchable-dropdowns (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-searchable-dropdowns)
✅ forum-search (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/forum-search)
✅ editor-stepping (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-stepping)
✅ image-uploader (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/image-uploader)
✅ forum-time-zones (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/forum-time-zones)
✅ remove-sprite-confirm (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/remove-sprite-confirm)
✅ progress-bar (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/progress-bar)
✅ remix-tree-button (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/remix-tree-button)
✅ bitmap-copy (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/bitmap-copy)
✅ full-signature (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/full-signature)
✅ 60fps (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/60fps)
✅ search-profile (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/search-profile)
✅ better-emojis (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/better-emojis)
✅ studio-tools (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/studio-tools)
✅ last-edit-tooltip (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/last-edit-tooltip)
✅ semicolon (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/semicolon)
✅ mouse-pos (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/mouse-pos)
✅ clones (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/clones)
✅ exact-count (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/exact-count)
✅ editor-messages (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-messages)
✅ pause (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/pause)
✅ resizable-comment-input (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/resizable-comment-input)
✅ animated-thumb (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/animated-thumb)
✅ project-info (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/project-info)
✅ confirm-actions (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/confirm-actions)
✅ feature-unshared (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/feature-unshared)
✅ block-switching (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/block-switching)
✅ my-ocular (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/my-ocular)
✅ color-picker (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/color-picker)
✅ onion-skinning (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/onion-skinning)
✅ data-category-tweaks-v2 (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/data-category-tweaks-v2)
✅ mute-project (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/mute-project)
✅ disable-auto-save (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/disable-auto-save)
✅ hide-flyout (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/hide-flyout)
✅ account-settings-capitalize (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/account-settings-capitalize)
✅ longer-wiwo (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/longer-wiwo)
✅ show-bbcode (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/show-bbcode)
✅ copy-message-link (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/copy-message-link)
✅ load-extensions (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/load-extensions)
✅ variable-manager (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/variable-manager)
✅ scratchstats (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/scratchstats)
✅ block-palette-icons (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/block-palette-icons)
✅ custom-block-shape (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/custom-block-shape)
✅ drag-drop (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/drag-drop)
✅ initialise-sprite-position (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/initialise-sprite-position)
✅ blocks2image (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/blocks2image)
✅ editor-sounds (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-sounds)
✅ transparent-orphans (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/transparent-orphans)
✅ gamepad (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/gamepad)
✅ remove-curved-stage-border (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/remove-curved-stage-border)
✅ fix-pasted-scripts (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/fix-pasted-scripts)
✅ better-img-uploads (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/better-img-uploads)
✅ ctrl-click-run-scripts (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/ctrl-click-run-scripts)
✅ expanding-search-bar (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/expanding-search-bar)
✅ disable-stage-drag-select (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/disable-stage-drag-select)
✅ studio-followers (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/studio-followers)
✅ forum-toolbar (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/forum-toolbar)
✅ items-per-row (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/items-per-row)
✅ ctrl-enter-post (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/ctrl-enter-post)
✅ no-script-bumping (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/no-script-bumping)
✅ user-id (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/user-id)
✅ hide-delete-button (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/hide-delete-button)
✅ name-backpack-items (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/name-backpack-items)
✅ forum-copy-code (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/forum-copy-code)
✅ paint-by-default (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/paint-by-default)
✅ hide-project-stats (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/hide-project-stats)
✅ editor-extra-keys (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-extra-keys)
✅ remix-button (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/remix-button)
✅ forum-id (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/forum-id)
✅ block-cherry-picking (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/block-cherry-picking)
✅ hide-new-variables (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/hide-new-variables)
✅ move-to-top-bottom (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/move-to-top-bottom)
✅ search-my-stuff (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/search-my-stuff)
✅ fullscreen (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/fullscreen)
✅ echo-effect (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/echo-effect)
✅ block-duplicate (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/block-duplicate)
✅ disable-paste-offset (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/disable-paste-offset)
✅ emoji-picker (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/emoji-picker)
✅ editor-comment-previews (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-comment-previews)
✅ swap-local-global (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/swap-local-global)
✅ fix-editor-comments (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/fix-editor-comments)
✅ tutorials-button (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/tutorials-button)
✅ block-count (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/block-count)
✅ script-snap (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/script-snap)
✅ custom-block-text (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/custom-block-text)
✅ search-sprites (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/search-sprites)
✅ hide-signatures (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/hide-signatures)
✅ disable-sprite-wobble (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/disable-sprite-wobble)
✅ forum-live-preview (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/forum-live-preview)
✅ default-costume-editor-color (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/default-costume-editor-color)
✅ remove-search-bar-autocomplete (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/remove-search-bar-autocomplete)
✅ zebra-striping (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/zebra-striping)
✅ find-bar (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/find-bar)
✅ jump-to-def (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/jump-to-def)
✅ dark-www (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/dark-www)
✅ scratchr2 (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/scratchr2)
✅ old-studio-layout (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/old-studio-layout)
✅ compact-messages (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/compact-messages)
✅ customize-avatar-border (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/customize-avatar-border)
✅ material-forum-editor-buttons (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/material-forum-editor-buttons)
✅ forum-quote-code-beautifier (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/forum-quote-code-beautifier)
✅ horizontal-mystuff-tabs (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/horizontal-mystuff-tabs)
✅ editor-dark-mode (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-dark-mode)
✅ editor-colored-context-menus (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-colored-context-menus)
✅ hide-stage (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/hide-stage)
✅ editor-stage-left (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-stage-left)
✅ editor-buttons-reverse-order (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-buttons-reverse-order)
✅ columns (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/columns)
✅ editor-theme3 (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-theme3)
🔄 (enable) editor-devtools (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-devtools)
🔄 (enable) true-youtube-links (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/true-youtube-links)
🔄 (enable) comments-linebreaks (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/comments-linebreaks)
❌ msg-count-badge (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/msg-count-badge)
❌ scratch-notifier (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/scratch-notifier)
❌ discuss-button (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/discuss-button)
❌ scratch-messaging (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/scratch-messaging)
❌ more-links (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/more-links)
❌ youtube-fullscreen (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/youtube-fullscreen)
❌ cloud-games (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/cloud-games)
❌ infinite-scroll (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/infinite-scroll)
❌ cat-blocks (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/cat-blocks)
❌ live-featured-project (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/live-featured-project)
❌ fix-uploaded-svgs (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/fix-uploaded-svgs)
❌ curator-link (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/curator-link)
❌ project-notes-tabs (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/project-notes-tabs)
❌ mediarecorder (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/mediarecorder)
❌ better-quoter (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/better-quoter)
❌ folders (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/folders)
❌ message-filters (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/message-filters)
❌ 2d-color-picker (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/2d-color-picker)
❌ dango-rain (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/dango-rain)
❌ custom-zoom (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/custom-zoom)
❌ redirect-mobile-forums (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/redirect-mobile-forums)
❌ debugger (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/debugger)
❌ default-project (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/default-project)
❌ turbowarp-player (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/turbowarp-player)
❌ editor-unshare-button (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/editor-unshare-button)
❌ hi-res-thumbnails (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/hi-res-thumbnails)
❌ color-inputs (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/color-inputs)
❌ sounds-duration (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/sounds-duration)
❌ one-click-studio-invites (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/one-click-studio-invites)
❌ scratchblocks (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/scratchblocks)
❌ better-featured-project (https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/better-featured-project)
--End of output--
73% (109/149) of addons are migrated to full refresh.
2% (3/149) of addons are partial refresh (enable).
0% (0/149) of addons are partial refresh (disable).
20% (31/149) of addons are not migrated to refresh.
```

**Comment by TheColaber** [comment]
Your code inspired me to make this code, which will create a todo list of what needs to be done:
<details>
<summary>code</summary>
<br>

```js
const addons = require("./addons/addons.json");
const categories = {
  themes: { full: [], onlyEnable: [], onlyDisable: [], no: [] },
  enabledByDefault: { full: [], onlyEnable: [], onlyDisable: [], no: [] },
  recommended: { full: [], onlyEnable: [], onlyDisable: [], no: [] },
  featured: { full: [], onlyEnable: [], onlyDisable: [], no: [] },
  other: { full: [], onlyEnable: [], onlyDisable: [], no: [] },
};
for (const addon of addons) {
  if (addon.startsWith("//")) continue;
  const { tags, dynamicEnable, dynamicDisable, enabledByDefault } = require(`./addons/${addon}/addon.json`);
  let type;
  if (dynamicEnable && dynamicDisable) {
    type = "full";
  } else if (dynamicEnable) {
    type = "onlyEnable";
  } else if (dynamicDisable) {
    type = "onlyDisable";
  } else {
    type = "no";
  }
  if (tags.includes("theme")) {
    categories.themes[type].push(addon);
  } else if (enabledByDefault) {
    categories.enabledByDefault[type].push(addon);
  } else if (tags.includes("recommended")) {
    categories.recommended[type].push(addon);
  } else if (tags.includes("featured")) {
    categories.featured[type].push(addon);
  } else {
    categories.other[type].push(addon);
  }
}
for (const type of Object.keys(categories)) {
  console.log(`## ${type}`);

  for (const addon of categories[type].full) {
    console.log(`- [x] [${addon}](https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/${addon})`);
  }
  for (const addon of categories[type].onlyEnable) {
    console.log(
      `- [ ] (only enable) [${addon}](https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/${addon})`
    );
  }
  for (const addon of categories[type].onlyDisable) {
    console.log(
      `- [ ] (only disable) [${addon}](https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/${addon})`
    );
  }
  for (const addon of categories[type].no) {
    console.log(`- [ ] [${addon}](https://github.com/ScratchAddons/ScratchAddons/tree/master/addons/${addon})`);
  }
}
```
</details>

**Comment by DNin01** [comment]
We can also [search](https://github.com/search?q=repo%3AScratchAddons%2FScratchAddons+path%3A%2F%5E*addon.json%2F+NOT+%28dynamicEnable+AND+dynamicDisable%29&type=code) to see addons that need either Dynamic Enable or Dynamic Disable.
